### ISSUE TITLE
[Feature Request] Ability to generate mouse events

### ISSUE THREAD

**Issue by aqua36** [david-janssen/kmonad#150]
I think it would be useful to be able to generate mouse events(mouse movement, mouse clicks etc.) from kmonad

**Potential use case:** It is conceivable that some people would find it useful to be able to emulate (parts of) a mouse from their keyboard

**My use case:** I own the Alienware AW958 mouse. It has six side buttons which under linux are detected as a keyboard that inputs the numbers 1..6 (*event9* in the attached picture)
![2020-12-08-08:44:31](https://user-images.githubusercontent.com/29358198/101460776-b3142a00-3942-11eb-82e7-7531c64488eb.png)

I would like to be able to bind those buttons to regular mouse buttons. At the moment I have used some workarounds (for example binding to XF86Forward and XF86Back to replicate the mouse next and mouse prev buttons)

There are probably other ways in which I can solve my particular problem, but I believe that, nevertheless, this feature would be a worthy addition to kmonad

Concluding, I wouldn't mind getting my hands dirty and trying to implement this myself. However, I would need (and highly appreciate) some guidance since I am only familiar with Haskell at a basic level and have no idea how keyboard and mouse events work. 

From what I have seen in [input-event-codes.h](https://github.com/torvalds/linux/blob/master/include/uapi/linux/input-event-codes.h) I think that mouse events (even button presses) are treated differently from keyboard button presses. I have found [this guide](https://www.kernel.org/doc/Documentation/input/input.txt) from the official kernel documentation that is probably the answer to how to read/interpret mouse events but I have not digged into it yet, due to time constraints.

**Comment by david-janssen** [comment]
Hey there,

That would indeed be a useful addition to KMonad. There are a few difficulties however:
- There is no support for this at the moment (i.e., you would need to go deep into the core to add mouse-events)
- It'll really complicate cross-platform support, every OS handles event-IO differently, so this would involve implementing the OS-specific code thrice

I'd be very happy to provide any amount of explanation on the structure of KMonad and brainstorm about how to implement such things, but I don't have the time/energy to be the main push behind this initiative.

So in summary: coooool idea, but know what you're getting into. I am definitely around to help, but you're going to have to be the driving force yourself :-)

**Comment by aqua36** [comment]
> There is no support for this at the moment (i.e., you would need to go deep into the core to add mouse-events)

Having only taken a look in the Haskell file where key events are defined I had such suspicions

> I'd be very happy to provide any amount of explanation on the structure of KMonad and brainstorm about how to implement such things, but I don't have the time/energy to be the main push behind this initiative.

Thank you for offering such help :) . Things for me are not much different at the moment, my free time is very limited, but it sounds like a fun side-project so I may be able to find some free time here and there to play around with this idea.

> It'll really complicate cross-platform support, every OS handles event-IO differently, so this would involve implementing the OS-specific code thrice

If I took this on, I would only be considering the linux case. Is not having feature parity for all platforms acceptable for this project?

**Comment by david-janssen** [comment]
> If I took this on, I would only be considering the linux case. Is not having feature parity for all platforms acceptable for this project?

Yeah, it's fine. I was just trying to surreptitiously rope you into doing some boring Windows dev-work ;-).

**Comment by aqua36** [comment]
Pretty sneaky... I like it :P

**Comment by molleweide** [comment]
for anyone using macos you can achieve mouse controll with hammerspoon 
eg. https://github.com/tweekmonster/hammerspoon-vimouse
l am going to try this and see how it feels. It could serve as inspiration for macos devs for the moment until/if this will be implement in kmonad later.

**Comment by molleweide** [comment]
yo I just found this project and it is written in haskell. I am not good enough yet to parse through it quickly but I'll share it.

https://github.com/unclechu/place-cursor-at

**Comment by molleweide** [comment]
there are also these written in C

https://github.com/SalmonChris/KeyboardMouse
https://github.com/Ph0enixKM/Mousepad

**Comment by molleweide** [comment]
linking to list of mouse nav projects https://github.com/kmonad/kmonad/issues/322

**Comment by gerhard-h** [comment]
Kmonad on Windows  stops most other mouse emulators (NeatMouse, Windows integrated, AHK with MouseMove) from working (plese reply if you found one that works.)

But autohotkey with `DllCall("SetCursorPos"` works, so here is the script I came up with:
Tap F23 to activate mouse keys
Click (u or o) to deactivate mouse keys
```
; ======================================================================
; Layer F23 mouse
; ======================================================================
MMove(L:=0, U:=0) {
	Mousespeed := 20
    VarSetCapacity(POINT, 8, 0)
    DllCall("GetCursorPos", "Ptr", &POINT)
    _X := NumGet(POINT, 0, "Int")
    _Y := NumGet(POINT, 4, "Int")
    DllCall("SetCursorPos", "int", _X + L*Mousespeed , "int", _Y + U*Mousespeed)
}

$F23::
	waitfor23 := true
	return


#If waitfor23

$u::
	MouseClick, left
	waitfor23 := false 
	return
$o::
	MouseClick, right
	waitfor23 := false
	return	
$k::
	MMove(0, 1)
	return
$i::
	MMove(0, -1)
	return
$j::
	MMove(-1, 0)
	return
$l::
	MMove(1, 0)
	return
#If
```

**Comment by jokober** [comment]
Everyone interested in this feature might also want to follow following feature request in the kanata repository: https://github.com/jtroo/kanata/issues/108

**Comment by al-jeyapal** [comment]
A possible workaround for now might be trying to leverage Xwindow mouse keys?
https://en.m.wikipedia.org/wiki/Mouse_keys

**Comment by manna-harbour** [comment]
> A possible workaround for now might be trying to leverage Xwindow mouse keys?
> https://en.m.wikipedia.org/wiki/Mouse_keys

- https://github.com/kmonad/kmonad/issues/322
- https://github.com/kmonad/kmonad/discussions/370
- https://github.com/manna-harbour/miryoku_kmonad#mouse-keys

**Comment by tusharhero** [comment]
> Layers
> A layer is a set of keymaps assigned to your keyboard's
> buttons. You can have as many layers on top of your base layer as you
> want. For instance, you can have your regular QWERTY layout, a
> Colemak/Dvorak layout, a numbers and symbols layer, a function keys
> layer, a layer for mouse navigation and system controls &#x2014; all in a
> 60% keyboard. When a particular layer is active, any keypress is
> interpreted according to the layout defined in that layer. With proper
> configurations, you can jump to a specific layer or switch to one for
> the next keypress, or do various other complex manipulations.

What does the readme mean by this?

**Comment by jokesper** [comment]
It's explained further in the tutorial. Though if you have more concrete questions this would also be helpful.
Please open a discussion instead, since this isn't related to the issue

**Comment by tusharhero** [comment]
I was referring to

> a layer for mouse navigation
